core/internal/runconsolelogs/linefile.go

```diff
@@ -0,0 +1,291 @@
+	origNextLine := f.nextLineNum
+
+	// Pop off old lines until the line we're modifying.
+	for ; f.nextLineNum > lines.FirstIndex(); f.nextLineNum-- {
+		line, err := file.PopLine()
+		if err != nil {
+			return err
+		}
+
+		oldLinesReversed = append(oldLinesReversed, line)
+	}
+
+	// Write back lines, replacing old lines as necessary.
+loop:
+	for {
+		var lineToWrite string
+
+		newLine, hadNewLine := lines.Get(f.nextLineNum)
+		switch {
+		// Use the new line if there is one.
+		case hadNewLine:
+			lineToWrite = newLine
+
+		// Use the old line if there is one.
+		case f.nextLineNum < origNextLine:
+			idx := origNextLine - 1 - f.nextLineNum
+			lineToWrite = oldLinesReversed[idx]
+
+		// If we're adding lines past the original number of lines in the
+		// file, append blanks for any unspecified lines.
+		case f.nextLineNum < lines.LastIndex():
+			lineToWrite = ""
+
+		// No more lines to write, stop.
+		default:
+			break loop
+		}
+
+		f.nextLineNum++
+
+		if err := file.AppendLine(lineToWrite); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Open opens the file for modification.
+func (f *lineFile) Open() (*lineFileIO, error) {
+	file, err := os.OpenFile(f.path, os.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file: %v", err)
+	}
+
+	info, err := file.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("failed to stat file: %v", err)
+	}
+
+	endOffset := info.Size()
+
+	return &lineFileIO{
+		f:           file,
+		bufStart:    endOffset,
+		buf:         make([]byte, 0),
+		nextLineNum: f.nextLineNum,
+		cursor:      endOffset,
+	}, nil
+}
+
+// Close flushes changes and closes the file handle.
+func (f *lineFileIO) Close() error {
+	if err := f.f.Truncate(f.bufStart + int64(len(f.buf))); err != nil {
+		return err
+	}
+
+	if err := f.f.Close(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AppendLine writes a line to the file.
+//
+// The line must not contain '\n'.
+func (f *lineFileIO) AppendLine(line string) error {
+	lineStart := f.bufStart + int64(len(f.buf))
+
+	// Truncate the file to the current position first.
+	if len(f.buf) > 0 {
+		if err := f.f.Truncate(lineStart); err != nil {
+			return err
+		}
+
+		f.buf = make([]byte, 0)
+	}
+
+	// Write the UTF-8 bytes of the line to the file with a '\n' at the end.
+	data := make([]byte, 0, len(line)+1)
+	data = append(data, line...)
+	data = append(data, '\n')
+
+	f.bufStart = lineStart + int64(len(data))
+	f.cursor = f.bufStart
+	f.nextLineNum += 1
+
+	_, err := f.f.WriteAt(data, lineStart)
+	return err
+}
+
+// PopLine drops the last file from the file and returns it.
+//
+// The returned line does not end with '\n'.
+//
+// Returns an error if there's an issue reading the file or if there are no
+// more lines.
+func (f *lineFileIO) PopLine() (string, error) {
+	if f.nextLineNum == 0 {
+		return "", errors.New("no lines remaining")
+	}
+
+	for {
+		err := f.cursorBack()
+		if err != nil {
+			return "", err
```

Comment:
do we close the file if we err?

Reply:
yes---in `lineFile.UpdateLines()`

the `Close()` is always near the `Open()`; since there's no `Open()` here it's not the responsibility of this method to `Close()`

Reply:
But I see what you're saying since it's not clear whether this could leave the reader in a corrupt state; in particular `cursor` won't be at the end of the file, so calling `PopLine()` again may not work as expected.

`Close()` will truncate the file but only based on `bufStart + len(buf)` which is unchanged on error.

Reply:
I updated `PopLine` to always reset `cursor` to the end of the file as well